### DISCUSSION TITLE
how to set_only_v6

### DISCUSSION THREAD

**Discussion by FH0** [tokio-rs/tokio#3175] [Q&A] [answered]
Assume the following two scenarios
- 0.0.0.0:80 is already bound by other application.
- The ipv6 address of the network interface changes frequently.

How to listen on :::80 via tokio?
```rust
async fn main() {
  // let v6 = TcpListener::bind(":::80", set_v6_only=true).await.unwrap();

  loop {
    v6.accept().await.unwrap();
  }
}
```

**Comment by taiki-e** [comment, marked as answer]
IIRC, neither `tokio::net::TcpListener` nor `std::net::TcpListener` supports this. You probably need to use [`socket2`](https://docs.rs/socket2/0.3/socket2/struct.Socket.html#method.set_only_v6) crate to create `std::net::TcpListener` and convert it to `tokio::net::TcpListener` with [`from_std`](https://docs.rs/tokio/0.3/tokio/net/struct.TcpListener.html#method.from_std) function.

**Comment by Darksonn** [comment]
What happens if you go through [`TcpSocket::new_v6`][1]?

[1]: https://docs.rs/tokio/0.3/tokio/net/struct.TcpSocket.html#method.new_v6

**Reply by FH0** [reply]
> What happens if you go through [`TcpSocket::new_v6`](https://docs.rs/tokio/0.3/tokio/net/struct.TcpSocket.html#method.new_v6)?

```rust
use tokio::net::TcpSocket;

use std::io;

#[tokio::main]
async fn main() -> io::Result<()> {
  let socket4 = TcpSocket::new_v4()?;
  socket4.bind("0.0.0.0:80".parse().unwrap())?;
  let socket6 = TcpSocket::new_v6()?;
  socket6.bind("[::]:80".parse().unwrap())?;

  socket4.listen(128)?;
  socket6.listen(128)?;

  std::thread::sleep(std::time::Duration::from_secs(1000));

  Ok(())
}
```

```
Error: Os { code: 98, kind: AddrInUse, message: "Address already in use" }
```
